File: backend/src/Service/MailScanService.php
<?php

namespace App\Service;

use App\Entity\Application;
use App\Entity\MailScan;
use App\Entity\StatusHistory;
use App\Entity\User;
use App\Enum\ApplicationStatus;
use App\Repository\ApplicationRepository;
use DateTimeImmutable;
use Doctrine\ORM\EntityManagerInterface;
use League\OAuth2\Client\Provider\Google;
use Symfony\Contracts\HttpClient\HttpClientInterface;
use Throwable;

class MailScanService
{
    private const GMAIL_API = 'https://gmail.googleapis.com/gmail/v1';
    private const MAX_RESULTS = 100;

    private const KEYWORDS_REFUSED = [
        // FR
        'nous avons le regret', 'avons le regret', 'regret de vous informer',
        'n\'avons pas retenu', 'n\'a pas été retenu', 'pas été retenue',
        'ne pas retenir', 'ne retenons pas', 'ne donnons pas suite',
        'ne donner pas suite', 'sans suite', 'avons décidé de ne pas poursuivre',
        'ne correspond pas à', 'votre profil ne correspond', 'candidature n\'a pas',
        'après étude de votre candidature', 'poursuivre avec d\'autres candidats',
        'nous ne sommes pas en mesure', 'n\'avons pas pu retenir', 'refus',
        // EN
        'unfortunately', 'regret to inform', 'not moving forward',
        'not selected', 'we will not be', 'decided not to proceed',
        'not a match', 'we have decided to', 'we won\'t be moving',
        'other candidates', 'position has been filled', 'no longer considering',
        'not the right fit', 'decided to move forward with other',
        'we regret', 'not successful',
    ];

    private const KEYWORDS_INTERVIEW = [
        // FR
        'entretien', 'rendez-vous', 'visioconférence', 'visio',
        'nous aimerions vous rencontrer', 'souhaitons échanger avec vous',
        'disponibilités', 'créneaux', 'prise de contact', 'appel téléphonique',
        'call téléphonique', 'vous rencontrer', 'échanger avec vous',
        'convocation', 'invitation à un entretien',
        // EN
        'interview', 'call with', 'meeting with', 'schedule a call',
        'schedule an interview', 'phone screen', 'phone call', 'video call',
        'would like to meet', 'set up a time', 'find a time', 'book a slot',
        'invite you to', 'next steps',
        // Tools
        'teams meeting', 'google meet', 'zoom', 'calendly',
    ];

    private const KEYWORDS_OFFER = [
        // FR
        'offre d\'emploi', 'proposition d\'embauche', 'nous vous proposons',
        'bienvenue dans', 'nous souhaitons vous accueillir',
        'lettre d\'engagement', 'heureux de vous proposer',
        'ravis de vous accueillir', 'rejoindre notre équipe',
        'offre de poste', 'vous rejoindre',
        // EN
        'pleased to offer', 'offer of employment', 'welcome to the team',
        'job offer', 'formal offer', 'offer letter', 'we would like to offer',
        'extend an offer', 'happy to offer', 'excited to offer',
        'we are delighted to offer',
    ];

    public function __construct(
        private readonly EntityManagerInterface $em,
        private readonly ApplicationRepository $applicationRepository,
        private readonly HttpClientInterface $httpClient,
        private readonly string $googleClientId,
        private readonly string $googleClientSecret,
    ) {
    }

    /** @return array{mailsAnalyzed: int, matchesFound: int, error?: string} */
    public function scanForUser(User $user): array
    {
        $accessToken = $this->refreshTokenIfNeeded($user);
        if (null === $accessToken) {
            $this->logAndFlush($user, 0, 0, 'error', 'Impossible de rafraîchir le token Gmail');

            return ['mailsAnalyzed' => 0, 'matchesFound' => 0, 'error' => 'Impossible de rafraîchir le token Gmail'];
        }

        $since = $user->getLastMailScanAt() ?? new DateTimeImmutable('-7 days');
        $applications = $this->applicationRepository->findBy(['user' => $user]);

        if (empty($applications)) {
            $user->setLastMailScanAt(new DateTimeImmutable());
            $this->em->flush();

            return ['mailsAnalyzed' => 0, 'matchesFound' => 0];
        }

        try {
            $messages = $this->fetchMessages($accessToken, $since);
        } catch (Throwable $e) {
            $this->logAndFlush($user, 0, 0, 'error', $e->getMessage());

            return ['mailsAnalyzed' => 0, 'matchesFound' => 0, 'error' => $e->getMessage()];
        }

        $matchesFound = 0;
        foreach ($messages as $message) {
            try {
                $detail = $this->fetchMessageDetail($accessToken, $message['id']);
                $matched = $this->processMessage($detail, $applications, $user);
                if ($matched) {
                    ++$matchesFound;
                }
            } catch (Throwable) {
                // skip malformed messages
            }
        }

        $user->setLastMailScanAt(new DateTimeImmutable());
        $this->logScan($user, count($messages), $matchesFound, 'success');
        $this->em->flush();

        return ['mailsAnalyzed' => count($messages), 'matchesFound' => $matchesFound];
    }

    // ─── Gmail API ────────────────────────────────────────────────────────────

    private function refreshTokenIfNeeded(User $user): ?string
    {
        $refreshToken = $user->getGmailRefreshToken();
        if (null === $refreshToken) {
            return $user->getGmailToken();
        }

        try {
            $provider = new Google([
                'clientId' => $this->googleClientId,
                'clientSecret' => $this->googleClientSecret,
                'redirectUri' => '',
            ]);

            $newToken = $provider->getAccessToken('refresh_token', [
                'refresh_token' => $refreshToken,
            ]);

            $user->setGmailToken($newToken->getToken());
            $this->em->flush();

            return $newToken->getToken();
        } catch (Throwable) {
            return $user->getGmailToken();
        }
    }

    /** @return array<array{id: string}> */
    private function fetchMessages(string $accessToken, DateTimeImmutable $since): array
    {
        $query = 'after:'.$since->format('Y/m/d');

        $response = $this->httpClient->request('GET', self::GMAIL_API.'/users/me/messages', [
            'headers' => ['Authorization' => 'Bearer '.$accessToken],
            'query' => ['q' => $query, 'maxResults' => self::MAX_RESULTS],
        ]);

        $data = $response->toArray(false);

        if (isset($data['error'])) {
            $msg = $data['error']['message'] ?? $data['error']['status'] ?? json_encode($data['error']);
            throw new \RuntimeException('Gmail API: '.$msg);
        }

        return $data['messages'] ?? [];
    }

    /** @return array{subject: string, from: string} */
    private function fetchMessageDetail(string $accessToken, string $messageId): array
    {
        $response = $this->httpClient->request('GET', self::GMAIL_API.'/users/me/messages/'.$messageId, [
            'headers' => ['Authorization' => 'Bearer '.$accessToken],
            'query' => [
                'format' => 'metadata',
                'metadataHeaders' => ['Subject', 'From'],
            ],
        ]);

        $data = $response->toArray();
        $headers = $data['payload']['headers'] ?? [];

        $subject = '';
        $from = '';
        foreach ($headers as $header) {
            if ('Subject' === $header['name']) {
                $subject = $header['value'];
            } elseif ('From' === $header['name']) {
                $from = $header['value'];
            }
        }

        return ['subject' => $subject, 'from' => $from, 'snippet' => $data['snippet'] ?? ''];
    }

    // ─── Matching ─────────────────────────────────────────────────────────────

    /** @param Application[] $applications */
    private function processMessage(array $message, array $applications, User $user): bool
    {
        $subject = mb_strtolower($message['subject']);
        $from = mb_strtolower($message['from']);
        $snippet = mb_strtolower($message['snippet']);
        $fromDomain = $this->extractDomain($from);

        foreach ($applications as $application) {
            if (!$this->matchesApplication($application, $subject, $from, $snippet, $fromDomain)) {
                continue;
            }

            $detectedStatus = $this->detectStatus($subject.' '.$snippet);
            if (null === $detectedStatus) {
                continue;
            }

            if ($detectedStatus === $application->getStatus()) {
                continue;
            }

            $this->updateApplicationStatus($application, $detectedStatus, $message['subject']);

            return true;
        }

        return false;
    }

    private function matchesApplication(
        Application $application,
        string $subject,
        string $from,
        string $snippet,
        string $fromDomain,
    ): bool {
        $normalizedCompany = $this->normalizeCompanyName($application->getCompanyName() ?? '');

        if ('' === $normalizedCompany) {
            return false;
        }

        // Check domain match
        if ('' !== $fromDomain && str_contains($fromDomain, $normalizedCompany)) {
            return true;
        }

        // Check company name in subject, from, or body snippet
        if (str_contains($subject, $normalizedCompany)
            || str_contains($from, $normalizedCompany)
            || str_contains($snippet, $normalizedCompany)) {
            return true;
        }

        // Check each word of the company name (for multi-word names)
        $words = array_filter(explode(' ', $normalizedCompany), fn (string $w) => mb_strlen($w) > 3);
        foreach ($words as $word) {
            if (str_contains($fromDomain, $word) || str_contains($subject, $word) || str_contains($snippet, $word)) {
                return true;
            }
        }

        return false;
    }

    private function detectStatus(string $subject): ?ApplicationStatus
    {
        $lower = mb_strtolower($subject);

        foreach (self::KEYWORDS_REFUSED as $kw) {
            if (str_contains($lower, $kw)) {
                return ApplicationStatus::REFUSED;
            }
        }

        foreach (self::KEYWORDS_OFFER as $kw) {
            if (str_contains($lower, $kw)) {
                return ApplicationStatus::OFFER;
            }
        }

        foreach (self::KEYWORDS_INTERVIEW as $kw) {
            if (str_contains($lower, $kw)) {
                return ApplicationStatus::INTERVIEW;
            }
        }

        return null;
    }

    private function updateApplicationStatus(
        Application $application,
        ApplicationStatus $newStatus,
        string $subjectNote,
    ): void {
        $previousStatus = $application->getStatus();
        $application->setStatus($newStatus);

        $history = new StatusHistory();
        $history->setApplication($application);
        $history->setPreviousStatus($previousStatus);
        $history->setNewStatus($newStatus);
        $history->setTrigger('auto_mail');
        $history->setNotes('Détecté depuis : '.$subjectNote);

        $this->em->persist($history);
    }

    private function logScan(
        User $user,
        int $mailsAnalyzed,
        int $matchesFound,
        string $status,
        ?string $errorMessage = null,
    ): void {
        $scan = new MailScan();
        $scan->setUser($user);
        $scan->setMailsAnalyzed($mailsAnalyzed);
        $scan->setMatchesFound($matchesFound);
        $scan->setStatus($status);
        $scan->setErrorMessage($errorMessage);

        $this->em->persist($scan);
    }

    private function logAndFlush(
        User $user,
        int $mailsAnalyzed,
        int $matchesFound,
        string $status,
        ?string $errorMessage = null,
    ): void {
        $this->logScan($user, $mailsAnalyzed, $matchesFound, $status, $errorMessage);
        $this->em->flush();
    }

    // ─── Utils ────────────────────────────────────────────────────────────────

    private function extractDomain(string $from): string
    {
        if (preg_match('/<([^>]+)>/', $from, $m)) {
            $email = $m[1];
        } elseif (str_contains($from, '@')) {
            $email = trim($from);
        } else {
            return '';
        }

        $parts = explode('@', $email);

        return mb_strtolower(trim(end($parts)));
    }

    private function normalizeCompanyName(string $name): string
    {
        $name = mb_strtolower($name);
        $name = (string) transliterator_transliterate('Any-Latin; Latin-ASCII', $name);
        // Remove legal suffixes
        $name = preg_replace('/\b(sas|sarl|sa|sasu|ltd|inc|gmbh|bv|nv|llc|corp|group|groupe)\b/', '', $name) ?? $name;
        // Keep only alphanumeric and spaces
        $name = preg_replace('/[^a-z0-9 ]/', '', $name) ?? $name;

        return trim((string) preg_replace('/\s+/', ' ', $name));
    }
}

File: backend-java/src/main/java/com/jobdesk/service/MailScanService.java
package com.jobdesk.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.jobdesk.domain.Application;
import com.jobdesk.domain.ApplicationStatus;
import com.jobdesk.domain.MailScan;
import com.jobdesk.domain.StatusHistory;
import com.jobdesk.domain.User;
import com.jobdesk.repository.ApplicationRepository;
import com.jobdesk.repository.MailScanRepository;
import com.jobdesk.repository.UserRepository;
import com.jobdesk.security.GoogleOAuthService;
import com.jobdesk.web.dto.ScanResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestClient;
import org.springframework.web.util.UriComponentsBuilder;

import java.text.Normalizer;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

/**
 * Scan des mails Gmail pour détecter les changements de statut des candidatures.
 * Portage de l'ancien {@code App\Service\MailScanService} (Symfony).
 */
@Service
public class MailScanService {

    private static final Logger log = LoggerFactory.getLogger(MailScanService.class);

    private static final String GMAIL_API = "https://gmail.googleapis.com/gmail/v1";
    private static final int MAX_RESULTS = 100;
    private static final DateTimeFormatter GMAIL_DATE = DateTimeFormatter.ofPattern("yyyy/MM/dd");

    private static final List<String> KEYWORDS_REFUSED = List.of(
            // FR
            "nous avons le regret", "avons le regret", "regret de vous informer",
            "n'avons pas retenu", "n'a pas été retenu", "pas été retenue",
            "ne pas retenir", "ne retenons pas", "ne donnons pas suite",
            "ne donner pas suite", "sans suite", "avons décidé de ne pas poursuivre",
            "ne correspond pas à", "votre profil ne correspond", "candidature n'a pas",
            "après étude de votre candidature", "poursuivre avec d'autres candidats",
            "nous ne sommes pas en mesure", "n'avons pas pu retenir", "refus",
            // EN
            "unfortunately", "regret to inform", "not moving forward",
            "not selected", "we will not be", "decided not to proceed",
            "not a match", "we have decided to", "we won't be moving",
            "other candidates", "position has been filled", "no longer considering",
            "not the right fit", "decided to move forward with other",
            "we regret", "not successful");

    private static final List<String> KEYWORDS_INTERVIEW = List.of(
            // FR
            "entretien", "rendez-vous", "visioconférence", "visio",
            "nous aimerions vous rencontrer", "souhaitons échanger avec vous",
            "disponibilités", "créneaux", "prise de contact", "appel téléphonique",
            "call téléphonique", "vous rencontrer", "échanger avec vous",
            "convocation", "invitation à un entretien",
            // EN
            "interview", "call with", "meeting with", "schedule a call",
            "schedule an interview", "phone screen", "phone call", "video call",
            "would like to meet", "set up a time", "find a time", "book a slot",
            "invite you to", "next steps",
            // Tools
            "teams meeting", "google meet", "zoom", "calendly");

    private static final List<String> KEYWORDS_OFFER = List.of(
            // FR
            "offre d'emploi", "proposition d'embauche", "nous vous proposons",
            "bienvenue dans", "nous souhaitons vous accueillir",
            "lettre d'engagement", "heureux de vous proposer",
            "ravis de vous accueillir", "rejoindre notre équipe",
            "offre de poste", "vous rejoindre",
            // EN
            "pleased to offer", "offer of employment", "welcome to the team",
            "job offer", "formal offer", "offer letter", "we would like to offer",
            "extend an offer", "happy to offer", "excited to offer",
            "we are delighted to offer");

    private final ApplicationRepository applicationRepository;
    private final UserRepository userRepository;
    private final MailScanRepository mailScanRepository;
    private final GoogleOAuthService google;
    private final RestClient http = RestClient.create();

    public MailScanService(ApplicationRepository applicationRepository,
                           UserRepository userRepository,
                           MailScanRepository mailScanRepository,
                           GoogleOAuthService google) {
        this.applicationRepository = applicationRepository;
        this.userRepository = userRepository;
        this.mailScanRepository = mailScanRepository;
        this.google = google;
    }

    @Transactional
    public ScanResult scanForUser(User user) {
        String accessToken = refreshTokenIfNeeded(user);
        if (accessToken == null) {
            logScan(user, 0, 0, "error", "Impossible de rafraîchir le token Gmail");
            return ScanResult.error("Impossible de rafraîchir le token Gmail");
        }

        LocalDateTime since = user.getLastMailScanAt() != null
                ? user.getLastMailScanAt()
                : LocalDateTime.now().minusDays(7);

        List<Application> applications = applicationRepository.findByUser(user);
        if (applications.isEmpty()) {
            user.setLastMailScanAt(LocalDateTime.now());
            userRepository.save(user);
            return ScanResult.ok(0, 0);
        }

        List<String> messageIds;
        try {
            messageIds = fetchMessageIds(accessToken, since);
        } catch (Exception e) {
            logScan(user, 0, 0, "error", e.getMessage());
            return ScanResult.error(e.getMessage());
        }

        int matches = 0;
        for (String messageId : messageIds) {
            try {
                MessageDetail detail = fetchMessageDetail(accessToken, messageId);
                if (processMessage(detail, applications)) {
                    matches++;
                }
            } catch (Exception e) {
                // skip malformed messages
                log.debug("Message {} ignoré: {}", messageId, e.getMessage());
            }
        }

        user.setLastMailScanAt(LocalDateTime.now());
        applicationRepository.saveAll(applications);
        userRepository.save(user);
        logScan(user, messageIds.size(), matches, "success", null);
        return ScanResult.ok(messageIds.size(), matches);
    }

    // ─── Gmail API ──────────────────────────────────────────────────────────────

    private String refreshTokenIfNeeded(User user) {
        String refreshToken = user.getGmailRefreshToken();
        if (refreshToken == null) {
            return user.getGmailToken();
        }
        try {
            var refreshed = google.refreshAccessToken(refreshToken);
            if (refreshed != null && refreshed.accessToken() != null) {
                user.setGmailToken(refreshed.accessToken());
                userRepository.save(user);
                return refreshed.accessToken();
            }
        } catch (Exception e) {
            log.warn("Refresh du token Gmail échoué: {}", e.getMessage());
        }
        return user.getGmailToken();
    }

    private List<String> fetchMessageIds(String accessToken, LocalDateTime since) {
        String uri = UriComponentsBuilder.fromUriString(GMAIL_API + "/users/me/messages")
                .queryParam("q", "after:" + since.toLocalDate().format(GMAIL_DATE))
                .queryParam("maxResults", MAX_RESULTS)
                .encode().build().toUriString();

        JsonNode body = http.get().uri(uri)
                .header("Authorization", "Bearer " + accessToken)
                .retrieve()
                .onStatus(status -> true, (req, res) -> { })
                .body(JsonNode.class);

        if (body == null) {
            return List.of();
        }
        if (body.has("error")) {
            throw new IllegalStateException("Gmail API: " + errorMessage(body.get("error")));
        }
        JsonNode messages = body.get("messages");
        if (messages == null || !messages.isArray()) {
            return List.of();
        }
        return messages.findValuesAsText("id");
    }

    private MessageDetail fetchMessageDetail(String accessToken, String messageId) {
        String uri = UriComponentsBuilder.fromUriString(GMAIL_API + "/users/me/messages/" + messageId)
                .queryParam("format", "metadata")
                .queryParam("metadataHeaders", "Subject")
                .queryParam("metadataHeaders", "From")
                .encode().build().toUriString();

        JsonNode body = http.get().uri(uri)
                .header("Authorization", "Bearer " + accessToken)
                .retrieve()
                .onStatus(status -> true, (req, res) -> { })
                .body(JsonNode.class);

        if (body == null || body.has("error")) {
            throw new IllegalStateException("Gmail API (detail)");
        }

        String subject = "";
        String from = "";
        JsonNode headers = body.path("payload").path("headers");
        if (headers.isArray()) {
            for (JsonNode h : headers) {
                String name = h.path("name").asText();
                if ("Subject".equals(name)) {
                    subject = h.path("value").asText("");
                } else if ("From".equals(name)) {
                    from = h.path("value").asText("");
                }
            }
        }
        String snippet = body.path("snippet").asText("");
        return new MessageDetail(subject, from, snippet);
    }

    // ─── Matching ─────────────────────────────────────────────────────────────

    private boolean processMessage(MessageDetail message, List<Application> applications) {
        String subject = message.subject().toLowerCase();
        String from = message.from().toLowerCase();
        String snippet = message.snippet().toLowerCase();
        String fromDomain = extractDomain(from);

        for (Application application : applications) {
            if (!matchesApplication(application, subject, from, snippet, fromDomain)) {
                continue;
            }
            ApplicationStatus detected = detectStatus(subject + " " + snippet);
            if (detected == null || detected == application.getStatus()) {
                continue;
            }
            updateApplicationStatus(application, detected, message.subject());
            return true;
        }
        return false;
    }

    private boolean matchesApplication(Application application, String subject, String from,
                                       String snippet, String fromDomain) {
        String company = normalizeCompanyName(application.getCompanyName() == null
                ? "" : application.getCompanyName());
        if (company.isEmpty()) {
            return false;
        }
        if (!fromDomain.isEmpty() && fromDomain.contains(company)) {
            return true;
        }
        if (subject.contains(company) || from.contains(company) || snippet.contains(company)) {
            return true;
        }
        for (String word : company.split(" ")) {
            if (word.length() > 3
                    && (fromDomain.contains(word) || subject.contains(word) || snippet.contains(word))) {
                return true;
            }
        }
        return false;
    }

    static ApplicationStatus detectStatus(String text) {
        String lower = text.toLowerCase();
        for (String kw : KEYWORDS_REFUSED) {
            if (lower.contains(kw)) {
                return ApplicationStatus.REFUSED;
            }
        }
        for (String kw : KEYWORDS_OFFER) {
            if (lower.contains(kw)) {
                return ApplicationStatus.OFFER;
            }
        }
        for (String kw : KEYWORDS_INTERVIEW) {
            if (lower.contains(kw)) {
                return ApplicationStatus.INTERVIEW;
            }
        }
        return null;
    }

    private void updateApplicationStatus(Application application, ApplicationStatus newStatus,
                                         String subjectNote) {
        ApplicationStatus previous = application.getStatus();
        application.setStatus(newStatus);

        StatusHistory history = new StatusHistory();
        history.setPreviousStatus(previous);
        history.setNewStatus(newStatus);
        history.setTrigger("auto_mail");
        history.setNotes("Détecté depuis : " + subjectNote);
        application.addStatusHistory(history);
    }

    private void logScan(User user, int mailsAnalyzed, int matchesFound, String status, String errorMessage) {
        MailScan scan = new MailScan();
        scan.setUser(user);
        scan.setMailsAnalyzed(mailsAnalyzed);
        scan.setMatchesFound(matchesFound);
        scan.setStatus(status);
        scan.setErrorMessage(errorMessage);
        mailScanRepository.save(scan);
    }

    // ─── Utils ────────────────────────────────────────────────────────────────

    private static String extractDomain(String from) {
        String email;
        int lt = from.indexOf('<');
        int gt = from.indexOf('>');
        if (lt >= 0 && gt > lt) {
            email = from.substring(lt + 1, gt);
        } else if (from.contains("@")) {
            email = from.trim();
        } else {
            return "";
        }
        int at = email.lastIndexOf('@');
        return at >= 0 ? email.substring(at + 1).trim().toLowerCase() : "";
    }

    /** Équivalent de transliterator Any-Latin; Latin-ASCII + nettoyage des suffixes légaux. */
    static String normalizeCompanyName(String name) {
        String n = name.toLowerCase();
        n = Normalizer.normalize(n, Normalizer.Form.NFD).replaceAll("\\p{M}", "");
        n = n.replaceAll("\\b(sas|sarl|sa|sasu|ltd|inc|gmbh|bv|nv|llc|corp|group|groupe)\\b", "");
        n = n.replaceAll("[^a-z0-9 ]", "");
        return n.replaceAll("\\s+", " ").trim();
    }

    private static String errorMessage(JsonNode error) {
        if (error.has("message")) {
            return error.get("message").asText();
        }
        if (error.has("status")) {
            return error.get("status").asText();
        }
        return error.toString();
    }

    private record MessageDetail(String subject, String from, String snippet) {
    }
}
